File: python/ransom_note.py
from collections import defaultdict


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        """time complexity: O(n)"""
        ransom_map: dict[str, int] = defaultdict(int)
        magazine_map: dict[str, int] = defaultdict(int)
        for c in magazine:
            magazine_map[c] += 1
        for c in ransomNote:
            ransom_map[c] += 1
            if c not in magazine_map or ransom_map[c] > magazine_map[c]:
                return False
        return True

    def canConstruct2(self, ransomNote: str, magazine: str) -> bool:
        r_count: list[int] = [0] * 26
        m_count: list[int] = [0] * 26
        for c in magazine:
            m_count[ord(c) - ord("a")] += 1
        for c in ransomNote:
            i: int = ord(c) - ord("a")
            r_count[i] += 1
            if m_count[i] == 0 or m_count[i] < r_count[i]:
                return False
        return True


def test_canConstruct_case_1():
    # arrange
    ransomNote: str = "a"
    magazine: str = "b"
    expected: bool = False

    # act
    actual = Solution().canConstruct(ransomNote, magazine)

    # assert
    assert expected == actual


def test_canConstruct_case_2():
    # arrange
    ransomNote: str = "aa"
    magazine: str = "ab"
    expected: bool = False

    # act
    actual = Solution().canConstruct(ransomNote, magazine)

    # assert
    assert expected == actual


def test_canConstruct_case_3():
    # arrange
    ransomNote: str = "aa"
    magazine: str = "aab"
    expected: bool = True

    # act
    actual = Solution().canConstruct(ransomNote, magazine)

    # assert
    assert expected == actual


def test_canConstruct_case_4():
    # arrange
    ransomNote: str = "aa"
    magazine: str = "aab"
    expected: bool = True

    # act
    actual = Solution().canConstruct2(ransomNote, magazine)

    # assert
    assert expected == actual

File: rust/src/code/ransom_note.rs
struct Solution;

impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        let mut r_count: Vec<i32> = vec![0; 26];
        let mut m_count: Vec<i32> = vec![0; 26];
        for c in magazine.chars() {
            m_count[c as usize - 'a' as usize] += 1;
        }
        for c in ransom_note.chars() {
            let i: usize = c as usize - 'a' as usize;
            r_count[i] += 1;
            if m_count[i] == 0 || r_count[i] > m_count[i] {
                return false;
            }
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_can_construct_case_1() {
        // arrange
        let ransom_note: String = "a".to_string();
        let magazine: String = "b".to_string();
        let expected: bool = false;

        // act
        let actual = Solution::can_construct(ransom_note, magazine);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_can_construct_case_2() {
        // arrange
        let ransom_note: String = "aa".to_string();
        let magazine: String = "ab".to_string();
        let expected: bool = false;

        // act
        let actual = Solution::can_construct(ransom_note, magazine);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_can_construct_case_3() {
        // arrange
        let ransom_note: String = "aa".to_string();
        let magazine: String = "aab".to_string();
        let expected: bool = true;

        // act
        let actual = Solution::can_construct(ransom_note, magazine);

        // assert
        assert_eq!(expected, actual);
    }
}
